mlir/lib/Dialect/IRDL/IR/IRDL.cpp

```diff
@@ -74,13 +74,62 @@ static void printSingleBlockRegion(OpAsmPrinter &p, Operation *op,
   if (!region.getBlocks().front().empty())
     p.printRegion(region);
 }
+static llvm::LogicalResult isValidName(llvm::StringRef in, mlir::Operation *loc,
+                                       const Twine &label) {
+  if (in.empty())
+    return loc->emitError("name of ") << label << " is empty";
+
+  bool allowUnderscore = false;
+  for (auto &elem : in) {
+    if (elem == '_') {
+      if (!allowUnderscore)
+        return loc->emitError("name of ")
+               << label << " should not contain leading or double underscores";
+    } else {
+      if (!isalnum(elem))
+        return loc->emitError("name of ")
+               << label
+               << " must contain only lowercase letters, digits and "
+                  "underscores";
+
+      if (llvm::isUpper(elem))
+        return loc->emitError("name of ")
+               << label << " should not contain uppercase letters";
+    }
+
+    allowUnderscore = elem != '_';
+  }
```

Comment:
I can't easily format on GitHub but I hope you get the gist.

```suggestion
  for (auto &elem : in) {
    if (elem == '_') {
      if (!allowUnderscore)
        return loc->emitError("name of ")
               << label << " should not contain leading or double underscores";
      allowUnderscore = false;
      continue;
    }
    
    if (!isalnum(elem))
        return loc->emitError("name of ")
               << label
               << " must contain only lowercase letters, digits and "
                  "underscores";

      if (llvm::isUpper(elem))
        return loc->emitError("name of ")
               << label << " should not contain uppercase letters";

    allowUnderscore = true:
  }
```